homeassistant/components/repetier/sensor.py

```diff
@@ -0,0 +1,336 @@
+
+SCAN_INTERVAL = timedelta(seconds=5)
+
+UPDATE_SIGNAL = 'repetier_update_signal'
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the available Repetier Server sensors."""
+    if discovery_info is None:
+        return
+
+    printer = discovery_info['printer']
+    sensor_type = discovery_info['sensor_type']
+    name = discovery_info['name']
+    data_key = discovery_info['data_key']
+
+    sensor_map = {
+        'bed_temperature': RepetierBedSensor,
+        'extruder_temperature': RepetierExtruderSensor,
+        'chamber_temperature': RepetierChamberSensor,
+        'current_state': RepetierStateSensor,
+        'current_job': RepetierJobSensor,
+        'time_remaining': RepetierRemainingSensor,
+        'time_elapsed': RepetierElapsedSensor,
+    }
+
+    sensor_class = sensor_map[sensor_type]
+    entity = sensor_class(printer, name, sensor_type, data_key)
+    add_entities([entity], True)
+
+
+class RepetierSensor(Entity):
+    """Class to create and populate a Repetier Sensor."""
+
+    def __init__(self, printer, name, sensor_type, data_key):
+        """Init new sensor."""
+        self._printer = printer
+        self._available = False
+        self._name = name
+        self._sensor_type = sensor_type
+        self._data_key = data_key
+        self._state = None
+        self._attributes = {}
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._available
+
+    @property
+    def device_state_attributes(self):
+        """Return sensor attributes."""
+        return self._attributes
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity, if any."""
+        return SENSOR_TYPES[self._sensor_type][1]
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend."""
+        return SENSOR_TYPES[self._sensor_type][2]
+
+    @callback
+    def update_callback(self):
+        """Get new data and update state."""
+        self.async_schedule_update_ha_state(True)
+
+    async def async_added_to_hass(self):
+        """Connect update callbacks."""
+        async_dispatcher_connect(
+            self.hass, UPDATE_SIGNAL, self.update_callback)
+
+
+class RepetierBedSensor(RepetierSensor):
+    """Class to create and populate a Repetier Bed Sensor."""
+
+    @property
+    def state(self):
+        """Return sensor state."""
+        if self._state is None:
+            return None
+        return round(self._state, 2)
+
+    def update(self):
+        """Update the sensor."""
+        self._printer.get_data()
+        if self._printer.heatedbeds is None:
+            self._available = False
+            return
+        if self._printer.state == "off":
+            self._available = False
+            return
+        self._available = True
```

Comment:
It would be cool if we can abstract the above logic into a general case for all sensor classes.

Reply:
?

Reply:
Look at the example code I posted. There I had the availability check in the common entity sensor class.

Reply:
But the different sensor type have different criteria for being available or not, how would you achive that in the RepetierSensor class?

Reply:
If we move all api interaction to a new class in the component, like I suggested, we can have a method that gets state from the state cache and returns `None` if the data is missing. We could use that return type to indicate unavailable.

```py
API_SENSOR_METHODS = {
    'bed_temperature': {
        'online' ['heatedbeds', 'state'],
        'state': {'heatedbeds': 'data_key'},
        'data_key': {
            'tempset': 'temp_set',
            'tempread': 'temp',
            'output': 'output'
       },
    },
    'extruder_temperature': ...
...
}


PrinterAPI:
    def __init__(self, client):
    self._client = client

    def get_data(self, printer_id, sensor_type, data_key):
        # get correct printer with printer_id
        printer.get_data()
        methods = API_SENSOR_METHODS[sensor_type]
        for prop in methods['online']:
            online = getattr(printer, prop)
            if online is None:
                return None

        data = {}
        for prop, attr in methods['state].items():
            prop_data = getattr(printer, prop)
            if attr == 'data_key':
                dk_props = methods['data_key']
                for dk_prop, dk_attr in dk_props.items():
                     data[dk_attr] = prop_data[data_key].dk_prop
            data[attr] = prop_data
        return data
```

Reply:
I give up. I don't get where you are going and I don't get how to implement the callbacks to get HA to add the missing sensors when they showed up - sorry.

Reply:
How would you call the PrinterAPI class? I'm missing the connection to the existing parts, I simply can't see how to implement this.

Reply:
I don't take into account multiple clients and api keys below.

In the component `setup` function:
```py
def setup(hass, config):
    ...
    client = pyrepetier.Repetier(...)
    api = PrinterAPI(client)
    hass.data[REPETIER_API] = api
    track_time_interval(hass, api.update, interval)
```
Define a method `update` in the api class where we look for new printers and sensors and update existing sensor via `dispatcher_send`.

`track_time_interval` will call the api update method every interval. The interval variable should be a timedelta type.

Reply:
If we still want to try to get this ready for merge, I suggest we simplify it and remove multiple api key option. You can also consider removing some sensor types.

It's ok to add these things later.

Reply:
I know of a few that are awaiting this to be merged, so I really try to get this right. But with documentation lacking information it is almost impossible to get to finish.

If ditching the possibility to have multiple api-keys/servers defined would make any of this easier - so be it.
Regarding sensor types - it wouldn't make much sense to remove any of them. State, job, remaning and elapsed are fine as is - it's the part about adding the temperature sensors as they become available that I don't get.
It did actually work in the initial version I made a PR on, but then you proposed all this callback stuff, of which I don't find much documented ø and you are referring to MySensors, which might be okay - but that one only makes the whole thing even more confusing.

Now you mention yet another term - `dispatcher_send` which I can't find documented anywhere, not even a Google on this gives anything related to Home Assistant.

And the PrinterAPI class - I really can't see what we need this for other than checking if a device are available or not :/